Exception Information

lp55 opened the pull request:
Adds exception information so the reporter on(exception) method has the exception message avaliable (for printing on screen)

claremacrae commented on include/boost/ut.hpp:
```diff
@@ -942,7 +942,11 @@ struct log {
 log(TMsg)->log<TMsg>;
 struct fatal_assertion {};
+#if defined(__cpp_exceptions)
+using exception = std::exception;
```
Some of the builds may need a #include to go with this - or maybe the std::runtime_error below - but see comments there too...

https://travis-ci.org/boost-experimental/ut/jobs/625157540#L1534

lp55 replied:
You're right. I thought about that later. I'll add it to the pull request

claremacrae commented on include/boost/ut.hpp:
```diff
@@ -1289,6 +1293,10 @@ class runner {
 #if defined(__cpp_exceptions)
       } catch (const events::fatal_assertion&) {
+      } catch (const std::exception& e) {
+        std::runtime_error errorMsg(e.what());
```
I'm unsure about this conversion of every std::exception type to std::runtime_error...

This is probably because I'm used to test frameworks saying things like:

"Unexpected exception of type MyCustomExceptionType occurred, with 'My informative explanation of the problem'"

The type of the exception can be really useful in narrowing down the cause, or location.

If it isn't possible to preserve the exception type here, I wonder whether saving the type-name of the exception in a std::string would be possible instead? It would likely be less confusing and more useful.

lp55 replied:
I used std::runtime_error because it has a constructor that takes a message and inherits from std::exception (which is basically what ApprovalException does for example) and it doesn't add a new dependency, and it also fulfils the initial goal of printing the exception message. To keep the original type of the exception would be trickier, and I haven't figured out a good way for that. If you use the throw<ExceptionType> function you can check for the type of the exception, but it's not the use case I had in mind (just printing the ApprovalTests message).

claremacrae commented on include/boost/ut.hpp:
```diff
@@ -1127,9 +1131,9 @@ class reporter {
+  auto on(events::exception exception) -> void {
     exception_ = true;
-    printer_ << "\n  " << printer_.colors().fail << "Unexpected exception!"
+    printer_ << "\n  " << printer_.colors().fail << exception.what()
```
Good idea to add the exception.what() here - I also think it would be worth keeping the original "Unexpected exception" text here too... not all exception messages may convey that an exception had occurred.... Even better if the exception type could be included as well.

This may be what Catch prints (it may not print the exception type):
https://github.com/catchorg/Catch2/blob/dd3d27de5787da803e031b561196905e59e681bc/include/reporters/catch_reporter_tap.hpp#L102

lp55 replied:
Actually if throw an exeption that doesn't override what(), it returns "Unknown exception", which seemed fine for me (similar enough to Unexpected exception!). But I changed that as well.

lp55 commented:
Hi, I've added tests and fixed some issues. codecov complains, but the diffs are minor. @krzysztof-jusiak is this ok? What more is needed for it to be merged?

lp55 commented:
@kris-jusiak / @krzysztof-jusiak 

I don't know why codecov is still complaining. I debugged the tests and the definitely hit the this code:

```cpp

#if defined(__cpp_exceptions)
      } catch (const events::fatal_assertion&) {
      } catch (const std::exception& e) {
        reporter_.on(events::exception{e.what()});
        active_exception_ = true;
      } catch (...) {
        reporter_.on(events::exception{"Unknown exception"});
        active_exception_ = true;
      }
#endif

```

with the new tests.
